### ISSUE TITLE
"Assignment" optimization

### ISSUE THREAD

**Issue by TurtlePU** [zkFold/zkfold-base#355]

**Comment by vlasin** [comment]
In the first case, it should be possible to optimize on the spot. While in the second case, we are working with an already built circuit.

**Comment by TurtlePU** [comment]
Yes it is -- we perform on-the-fly evaluation and optimize if new variable depends linearly on some other variable, but we do not perform "backwards" optimization from the linear constraint. For now, `optimize` only deals with "backwards" constant optimization by finding linear constraints with a single variable.
